mist-core/src/main/java/edu/snu/mist/core/master/ApplicationAwareQueryAllocationManager.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2018 Seoul National University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.mist.core.master;
+
+import edu.snu.mist.core.parameters.OverloadedTaskThreshold;
+import edu.snu.mist.formats.avro.IPAddress;
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * The app-allocation manager which allocates queries in application-aware way.
+ */
+public final class ApplicationAwareQueryAllocationManager implements QueryAllocationManager {
+
+  /**
+   * The map which maintains the app-task list information.
+   */
+  private final ConcurrentMap<String, List<IPAddress>> appTaskListMap;
+
+  /**
+   * The lock used for preventing race condition when modifying app task list.
+   */
+  private final ConcurrentMap<String, Lock> modifyTaskListLockMap;
+
+  /**
+   * The map which maintains task info.
+   */
+  private final ConcurrentMap<IPAddress, TaskInfo> taskInfoMap;
+
+  /**
+   * The threshold for determining overloaded task.
+   */
+  private final double overloadedTaskThreshold;
+
+  @Inject
+  private ApplicationAwareQueryAllocationManager(
+      @Parameter(OverloadedTaskThreshold.class) final double overloadedTaskThreshold
+  ) {
+    this.appTaskListMap = new ConcurrentHashMap<>();
+    this.modifyTaskListLockMap = new ConcurrentHashMap<>();
+    this.taskInfoMap = new ConcurrentHashMap<>();
+    this.overloadedTaskThreshold = overloadedTaskThreshold;
+  }
+
+  // TODO: [MIST-519] Consider query reallocation.
+  @Override
+  public synchronized IPAddress getAllocatedTask(final String appId) {
```

Comment:
If we assign applications to the minimum loaded task, many applications could be assigned to the same task at the same time. We need to prevent this situation. Could you please create an issue for it?

Reply:
Created on #1012.